src/main/scala/inox/ast/ExprOps.scala

```diff
@@ -65,30 +65,80 @@ trait ExprOps extends GenTreeOps {
   }
 
   /** Freshens all local variables
-    * 
+    *
     * Note that we don't freshen choose ids as these are considered global
     * and used to lookup their images within models!
     */
-  def freshenLocals(expr: Expr, freshenChooses: Boolean = false): Expr = {
-    val subst: MutableMap[Variable, Variable] = MutableMap.empty
-    variablesOf(expr).foreach(v => subst(v) = v)
+  class Freshener(freshenChooses: Boolean = false) extends SelfTransformer {
```

Comment:
Could you make this class and the two vals below protected to help indicate these are override points?

Also, you probably shouldn't use a default value for `freshenChooses` here to make sure sub-classes correctly propagate the constructor parameter.